### PR TITLE
Fix event types and .once this binding

### PR DESCRIPTION
**Pull request by jiayihu**
- Made event type definitions interfaces to allow again declaration merging
- Fixed the types for the `on, once, off` event registering. It was particularly not working for registering with an object of event handlers, since the type of `EventRegistryObject` was `Record<K, TEventCallback<E>>`, meaning that **every** event has the same callback type, whereas each event has its own.
- Fixed a bug with `obj.once("event", function() { console.log(this) })` where `this === undefined`, whereas `obj.on("event", function() { console.log(this) })` would correctly have `this === object`

### CODE REVIEW

File: src/Observable.ts
Comment (ShaMan123) [inline src/Observable.ts:3]: should E extend `Record<string, any>`?
Then `K` should always be string and no symbol etc.

File: src/Observable.ts
Comment (ShaMan123) [inline src/Observable.ts:33]: Why did you remove this overload?
The base signature is not exposed by TS to the consumer, is it?

File: src/Observable.ts
Comment (ShaMan123) [inline src/Observable.ts:44]: I dislike this syntax.
`event` is a confusing name in this context, I also prefer inlining it of possible.
Does `K` become a string after applying my first comment? If so it resolves all this.

File: src/Observable.ts
Comment (ShaMan123) [inline src/Observable.ts:93]: maybe we should give it a name for debugging
`onceHandler`?

**Review by ShaMan123 — COMMENTED** [review]
Looks good, see comments
Could yuo add a test for the once handler, so we cover it?

**Comment by ShaMan123** [comment]
Also, what do you think about a TS test suite? #8392

File: src/Observable.ts
Comment (jiayihu) [reply src/Observable.ts:3]: > should E extend Record<string, any>?

I tried it but it doesn't seem to help to restrict the key type to `string`. In turn it would force `EventSpec` to be a `Record` it self, which is okay, I just didn't do it to avoid in turn affecting maybe other types as well. But if you're okay with it, I can try.

File: src/Observable.ts
Comment (jiayihu) [reply src/Observable.ts:33]: It looks removed but it isn't, look at the final code on the right
Screenshot 2023-07-29 at 13 00 54

File: src/Observable.ts
Comment (jiayihu) [reply src/Observable.ts:44]: > Does K become a string after applying my first comment? If so it resolves all this.

When I tried yesterday, no. I didn't investigate further though. 

Anyway it would still be:

```ts
this.on<K, EventSpec[K]>(eventName as K, arg0[eventName as K]!);
```

Are you okay with that? For me all that typing makes the code less readable than `this.on<K, EventSpec[K]>(event, arg0[event]!)` but it's up to you.

File: src/Observable.ts
Comment (jiayihu) [reply src/Observable.ts:3]: Okay I've tried it and it still gives this error

Screenshot 2023-07-29 at 13 09 09

Using `Object.entries` fixes the issue, probably `for (key in obj)` is hardcoded to have type `string | number | symbol` because you're looping over all the keys, even non-own properties.

However, `extends Record<string, any>` brought to `CommonMethods<CanvasEvents>` giving error because `CanvasEvents` cannot be assigned to `Record<string, any>`. It works only with `type`, not `interface`. But we need `interface` for declaration merging.

See:
- https://github.com/microsoft/TypeScript/issues/15300
- https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsAShA5gSwM7AE4gA8MUEAHsEgCbZRoDGA9nlYbnpougDRSICuAWwBGEPAD5xUALxQA3lADaAaSicoAawghGAM1gBdAFx8hovFAC+AKGuhIsBMgDKkejLoQmLNvk49TETFxa05KPF0AQ3poeCRgVy8AJnlrKHSoSJMk6xs7cGgAQQ845DQsdiJShLcQ+2gAIRKnVAwcfCqWxPok8SA

Comment (ShaMan123) [reply src/Observable.ts]: Does [this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsAShA5gSwM7AE4gA8MUEAHsEgCbZRoDGA9nlYbnpougDRSICuAWwBGEPAD5xUALxQA3lADaAaSicoAawghGAM1gBdAFx8hovFAC+AKGuhIsBMgDKkejLoQmLNvk49TETFxa05KPF0AQ3poeCRgVy8AJlIKalo4lzd5ayg8qEiTJOsbO3BoAEEPTNQMHHwiGsT6EPtoACFqp1qsdkbu5qTxIA) solve the issue?

I didn't fully understand the interface limitation. Didn't read it thoroughly.
This is when I say ohh TS... why???

Comment (ShaMan123) [reply src/Observable.ts]: I am ok using Object.entries
Very annoying the for in loop wrong stqtic types

Comment (ShaMan123) [reply src/Observable.ts]: Another question about declaration merging
Why not pass an augmented type to the class? That was my intention we I wrote it. I didn't think of declaratiin mergign because I don't use it so much.
I am used to generics

Comment (ShaMan123) [reply src/Observable.ts]: So if i understand correctly using types is better for maintainers because TS is simpler to use and interfaces for devs/consumers for declarstiin merging but that takes a toll from maintaining?

Comment (ShaMan123) [reply src/Observable.ts]: > I am ok using Object.entries Very annoying the for in loop wrong static types Just for discussion, do you know of a perf diff between them?

Or we just `@ts-expect-error`

**Review by ShaMan123 — APPROVED** [review]
I noticed this

https://github.com/jiayihu/fabric.js/blob/35a27e37606db22fb857b05891d55e422cb99a8d/src/Observable.ts#L27-L30

https://github.com/jiayihu/fabric.js/blob/35a27e37606db22fb857b05891d55e422cb99a8d/src/Observable.ts#L70-L73

Or on master:

https://github.com/fabricjs/fabric.js/blob/d2c7b897664c6821339d142b41e7a8023bed965f/src/Observable.ts#L28-L31

https://github.com/fabricjs/fabric.js/blob/d2c7b897664c6821339d142b41e7a8023bed965f/src/Observable.ts#L73-L76

Why is it there? I am to blame... Maybe the prev type needed it.
Seems that it should be removed, right?

File: src/Observable.ts
Comment (jiayihu) [reply src/Observable.ts]: > Does [this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsAShA5gSwM7AE4gA8MUEAHsEgCbZRoDGA9nlYbnpougDRSICuAWwBGEPAD5xUALxQA3lADaAaSicoAawghGAM1gBdAFx8hovFAC+AKGuhIsBMgDKkejLoQmLNvk49TETFxa05KPF0AQ3poeCRgVy8AJlIKalo4lzd5ayg8qEiTJOsbO3BoAEEPTNQMHHwiGsT6EPtoACFqp1qsdkbu5qTxIA) solve the issue?

Not really, because at that point you can do `const value: EventSpec2 = { a: 2, b: 3 }`, e.g. add extra properties without error. And that's exactly why TS gives the error in my example, because technically you're saying that something with "well-defined properties" is assignable to a type where "any string index can be used", which is unsafe.

> Just for discussion, do you know of a perf diff between them?

Screenshot 2023-07-30 at 12 43 07

There is definitely an overhead in using functions, but not significant. Also not relevant for an object of event handlers, whose number of entries is only a few and the code is not in a critical hot path (rendering).

> Why not pass an augmented type to the class? That was my intention when I wrote it. I didn't think of declaration merging because I don't use it so much. I am used to generics

Declaration merging is different than generics. Generic is for a specific instance, a declaration merging extends an existing interface wherever it's used. It can also significantly harder for the lib to maintain good generics. For instance the following would need to be fixed as the `EventSpec` generic is currently not used:

Screenshot 2023-07-30 at 12 45 29

But if you fix everything and make it work, it's definitely a doable alternative. At that point not sure if there are still other use cases where declaration merging for canvas/object events might be helpful.

**Comment by jiayihu** [comment]
> Why is it there? I am to blame... Maybe the prev type needed it.
Seems that it should be removed, right?

I don't know but I noticed it to be weird indeed. I can remove it if you want, no idea why it was there though.

**Comment by ShaMan123** [comment]
> > Why is it there? I am to blame... Maybe the prev type needed it.
> > Seems that it should be removed, right?
> 
> I don't know but I noticed it to be weird indeed. I can remove it if you want, no idea why it was there though.

lets remove the wierd artifact

**Comment by jiayihu** [comment]
Done

**Review by ShaMan123 — APPROVED** [review]
I am good with this PR, it also fixes autocomplete of `.on({ ... })`
Declaration merging, interface vs. type, poses a problem that we should decide about soon.
@asturur do you accept it for 6.0?

**Review by asturur — DISMISSED** [review]

**Pull request merged**
